src/ed25519/mod.rs

```diff
@@ -0,0 +1,92 @@
+use crate::{
+    constants::ED25519_TAG,
+    crypto_tools::{message_digest::MessageDigest, rng},
+    sdk::api::{BytesVec, TofnFatal, TofnResult},
+};
+use ed25519_dalek::{Signature, Signer, SigningKey, VerifyingKey, PUBLIC_KEY_LENGTH};
+
+#[derive(Debug)]
+pub struct KeyPair(SigningKey);
+
+impl KeyPair {
+    pub fn encoded_verifying_key(&self) -> [u8; PUBLIC_KEY_LENGTH] {
+        *self.0.verifying_key().as_bytes()
+    }
+}
+
+pub fn keygen(
+    secret_recovery_key: &rng::SecretRecoveryKey,
+    session_nonce: &[u8],
+) -> TofnResult<KeyPair> {
+    let mut rng = rng::rng_seed_ecdsa_signing_key(
+        ED25519_TAG,
+        KEYGEN_TAG,
+        secret_recovery_key,
+        session_nonce,
+    )?;
+
+    let signing_key = SigningKey::generate(&mut rng);
+
+    Ok(KeyPair(signing_key))
+}
+
+/// Returns the 64 bytes signature containing the (R, s) components.
```

Comment:
It might've been nice to return an ASN.1 DER encoded signature here, but that doesn't appear to be supported by the underlying crates.

There's an implementation of the DER encoding in https://github.com/mobilecoinfoundation/mobilecoin/blob/master/crypto/keys/src/ed25519.rs#L43-L112, I can grab it from there if desirable.

Reply:
> but that doesn't appear to be supported by the underlying crates.

I'd be happy to add an implementation to the @RustCrypto `ed25519` crate, or potentially the `signature` crate, given the generic nature of this particular ASN.1 schema (it appears to be a small bit of X.509 described in RFC5912, and we already have precedent for something like that in the `spki` crate).

We already implement PKCS#8/SPKI there, but an encoding for the signature itself is a missing detail.

Reply:
It looks like we already have an implementation here: https://github.com/RustCrypto/formats/blob/6e60929b2903b64d4d4fd5d3f2e705680352519d/x509-ocsp/src/lib.rs#L102-L107

Unfortunately the `certs` field requires pulling in all of X.509.

Reply:
Interesting. Yeah, it would be nice if it made it into `ed25519`.

I am still not sure how the `Signature` struct from https://github.com/RustCrypto/formats/blob/6e60929b2903b64d4d4fd5d3f2e705680352519d/x509-ocsp/src/lib.rs#L102-L107 can be used to produce the DER encoded bytes in its current state. Unless I'm missing something it is still missing the impls to actually do the encoding.

Reply:
It's using custom derive provided by [`der_derive`](https://crates.io/crates/der_derive) which impls the relevant traits via `derive(Sequence)` and see also the `asn1` attribute on `certs`.

Now I'm noticing we never actually published this crate or I could link you the rustdoc on docs.rs, but here's a similar type with derived encoding traits:

https://docs.rs/x509-cert/latest/x509_cert/certificate/struct.TbsCertificateInner.html

Reply:
Oh, my bad, I missed the derive.

Reply:
This was helpful, thank you. I changed it to return the ASN1 DER encoded bytes. I also printed them and compared to the hardcoded header in that MobileCoin code linked above and it matched :)